core_impl/src/main/java/com/google/instrumentation/stats/StatsSerializer.java

```diff
@@ -31,52 +29,81 @@
  * Native implementation {@link StatsContext} serialization.
  */
 final class StatsSerializer {
-  //  * tag_metadata is one byte and is for encoding metadata about the tag. The
-  //    low 2 bits of this byte are used to encode the type of the value bytes.
-  //    The high 6 bits are reserved for future use. The value_bytes type is
-  //    encoded as:
-  //    00 (value 0): string (UTF-8) encoding
-  //    01 (value 1): integer (varint int64 encoding).
-  //    10 (value 2): boolean format.
-  //    11 (value 3): reserved for future use.
+
+  // This describes the encoding of stats context information (tags)
+  // for passing across RPC's.
+  // Tags are encoded in single byte sequence. The format is:
+  // <version_id><encoded_tags>
+  //   <version_id> == a single byte, value 0
+  //   <encoded_tags> == (<tag_field_id><tag_encoding>)*
+  //     <tag_field_id> == a single byte, value 0 (string tag), 1 (int tag),
+  //                       2 (true bool tag), 3 (false bool flag)
+  //     <tag_encoding> (tag_field_id == 0) ==
+  //       <tag_key_len><tag_key><tag_val_len><tag_val>
+  //         <tag_key_len> == varint encoded integer
+  //         <tag_key> == tag_key_len bytes comprising tag key name
+  //         <tag_val_len> == varint encoded integer
+  //         <tag_val> == tag_val_len bytes comprising UTF-8 string
+  //     <tag_encoding> (tag_field_id == 1) ==
+  //       <tag_key_len><tag_key><int_tag_val>
+  //         <tag_key_len> == varint encoded integer
+  //         <tag_key> == tag_key_len bytes comprising tag key name
+  //         <int_tag_value> == 8 bytes, little-endian integer
+  //     <tag_encoding> (tag_field_id == 2 || tag_field_id == 3) ==
+  //       <tag_key_len><tag_key>
+  //         <tag_key_len> == varint encoded integer
+  //         <tag_key> == tag_key_len bytes comprising tag key name
```

Comment:
The only requirement is that newly added ids must be added at the end (appended) because the old implementations don't know about the format of that field. For the initial known tags no restriction for that.

Also you may want to look at PropagationUtil (from trace) on how to structure/format the comment to be consistent between these.